Shadow system 3 / 6.

realitix opened the pull request:
This commit refactor shadow system to be more consistent.
Moreover, it adds BaseShadowSystem which is the base for systems.

See https://github.com/libgdx/libgdx/pull/3347

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+		this.lightFilter = lightFilter;
+		frameBuffers = new FrameBuffer[getPassQuantity()];
+		passShaderProviders = new ShaderProvider[getPassQuantity()];
+		init();
```
Calling a protected method in the constructor of the base class is not a good idea ;)

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+	}
+
+	/** First pass number */
+	public static final int PASS_1 = 0;
```
Assuming zero based indexing, what happened to `PASS_0`?

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+	protected int currentPointSide;
+	protected PointLightProperties currentPointProperties;
+	/** Tmp variables */
+	protected Vector3 tmpV3 = new Vector3();
```
Why is this here? It doesnt seem to be used. Also why is this protected? You typically want to keep the scope of tmp instances small to avoid problems because you dont know where they are used.

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+		Gdx.gl.glScissor(r.x, r.y, r.width, r.height);
+		region.setRegion(r.x, r.y, r.width, r.height);
+
+		return tmpV2.set(r.width, r.height);
```
Better add an `out` argument to remove the need of a tmp vector.

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+
+	@Override
+	public ShaderProvider getPassShaderProvider (int n) {
+		if (n >= passShaderProviders.length)
```
I dont think that this check is needed here (otherwise it will throw an index out of bounds exception, which says the same). https://github.com/libgdx/libgdx/wiki/Contributing#performance-considerations

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+		this.allocator = allocator;
+		this.directionalAnalyzer = directionalAnalyzer;
+		this.lightFilter = lightFilter;
+		frameBuffers = new FrameBuffer[getPassQuantity()];
```
You are calling an abstract protected method in the constructor to initialize an array. That will probably go wrong on any implementation other than the basic one.

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+				camera.direction.set(cubemapSide.direction);
+				camera.up.set(cubemapSide.up);
+				camera.near = 0.1f;
+				camera.far = 1000;
```
Probably better to use `1f` and `100f` or so.

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+	}
+
+	@Override
+	public void begin (Camera camera, Iterable<RenderableProvider> renderableProviders) {
```
this will not work for an api, please use `<T extends RenderableProvider>`. Like so: https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/g3d/ModelBatch.java#L253. And/or dont use generics and cast it instead.

xoppa commented on tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java:
```diff
@@ -0,0 +1,468 @@
+		currentPass = n;
+
+		spotCameraIterator = spotCameras.iterator();
+		spotCameraIterator.reset();
```
Calling the `iterator` method always resets it, so you only need one of these two methods.

realitix commented:
Hi Xoppa, thanks for the review, here my update in the order:

1 - Why is it bad to call protected method in constructor ?
init() should be overriden. What is the good practice ?

2 - PASS_0 should not exist. PASS_1 is just a label to access the first pass framebuffer.
I can rename it to PASS_0 if you want.

3 - tmpV3 is deleted

4 - processViewport() now update camera itself, so I removed tmpV2.

5 - I removed the check in getPassShaderProvider()

6 - Why is it a bad to call an abstract protected method in the constructor ?
It must return at least 1 (comments). What is the good practice ?

7 - I use 1, 100 for camera default values.

8 - I added <T extends RenderableProvider> but I'm not confident.
I had to cast it, is it good ?

9 - I removed reset() from iterator.

davebaol commented:
> 1 - Why is it bad to call protected method in constructor ?
> init() should be overriden. What is the good practice ?

http://stackoverflow.com/a/3404369

realitix commented:
Hi Xoppa, I updated the BaseShadowSystem according to your comments, I hope all is good this time!
I removed all reference to first pass in BaseShadowSystem. I created a new class named FirstPassBaseShadowSystem.

realitix commented:
I forgot Scissor padding !

realitix commented:
I added scissor padding
